ACPs/194-streaming-asynchronous-execution/README.md

```diff
@@ -118,49 +125,123 @@ ACP-103 introduced the following variables for calculating the gas price:
+
+### Queue size
+
+The constant time delay between execution and settlement is defined as $\tau$ seconds.
+
+The maximum allowed size of the execution queue is defined as:
+
+$$
+\omega ~:= R \cdot \tau \cdot \lambda
+$$
+
+Any block that could cause the total sum of gas limits for transactions in the execution queue to exceed $\omega$ MUST be considered invalid.
+
+### Block executor
+
+During the activation of SAE, the block executor's timestamp, $t_e$, is initialised to the timestamp of the last accepted block.
 
-$$x := x + \frac{g \cdot (p-1)}{p}$$
+Prior to executing a block which includes timestamp $t_b$, the executor's timestamp and excess is updated:
 
-and the gas price is updated immediately after execution, according to the same formula.
-When the transaction queue is drained and a chunk has a gas _surplus_ of $0 \le s \le R$, the excess is decreased:
+$$
+\begin{align}
+\Delta{t} &~:= \max\left(t_e, t_b\right) - t_e \\
+t_e &~:= t_e + \Delta{t} \\
+x &~:= \max\left(x - T \cdot \Delta{t}, 0\right) \\
+\end{align}
+$$
 
-$$x := \max\left(x-\frac{s}{p}, 0\right)$$
+The block is then executed with the gas price calculated by the current value of $x$.
 
-and the gas price is updated before the next chunk.
-With the current $p = 2$, half of all gas consumed is added to excess while half of all surplus is deducted.
+After executing a block which was charged $g_C$ gas, the executor's timestamp and excess is updated:
+
+$$
+\begin{align}
+\Delta{t} &~:= \frac{g_C}{R} \\
+t_e &~:= t_e + \Delta{t} \\
+x &~:= x + \Delta{t} \cdot (R - T) \\
+\end{align}
+$$
 
 > [!NOTE]
-> The ACP-103 concept of a leaky gas bucket with capacity $C$ is an emergent phenomenon in SAE.
+> The update rule here assumes that $t_e$ is a timestamp that can track the passage of time both by gas and by wall-clock time. $\frac{g_C}{R}$ MUST NOT be simply rounded. Rather, the gas accumulation should be left as a fraction.
 
-### JSON RPC methods
+$t_e$ is now this block's execution timestamp.
 
-Although asynchronous execution decouples the transactions and receipts recorded by a specific block, APIs MUST NOT alter their behavior to mirror this.
-In particular, the API method `eth_getBlockReceipts` MUST return the receipts corresponding to the block's transactions, not the receipts settled in the block.
+### Handling gas target changes
 
-To achieve this, we first define the _executed_ and _settled_ states for blocks.
-A block is initially _accepted_ by consensus, enqueuing its tranche of transactions.
-Once the last transaction in the tranche has been executed, the block itself is considered to be executed.
-When the same transaction is settled then so too is the block considered settled.
+When a block is produced which modifies $T$, both the consensus thread and the execution thread will update to the modified $T$ after their own handling of the block.
```

Comment:
```suggestion
When a block is produced that modifies $T$, both the consensus thread and the execution thread will update to the modified $T$ after their own handling of the block.
```

Also, why _after_? Is that for simplicity, or is it just in keeping with the current approach from ACP-176?

Reply:
It was picked for simplicity in ACP-176, and I think it is still the most simple design. I would have to look back into the code tbh, but I remember avoiding to need to have access to grand-parent blocks during some codepaths by applying the change after the block.

It also makes it so that block builders can't (are less likely to) act selfishly. (If they don't want the block sizes to increase but will increase their own block sizes to include more transactions, then we get into some weird situations).